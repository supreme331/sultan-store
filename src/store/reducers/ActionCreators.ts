import {createAsyncThunk} from "@reduxjs/toolkit";
import {fetchProductsImitation} from "../../utils/utils";

export const fetchProductItems = createAsyncThunk('catalog/fetchProductItems', async (_, thunkAPI) => {
    try {
        return await fetchProductsImitation();
    } catch (e) {
        return thunkAPI.rejectWithValue('Ошибка при получении товаров.')
    }
})