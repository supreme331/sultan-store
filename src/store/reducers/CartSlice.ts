import {createSlice, PayloadAction} from "@reduxjs/toolkit";
import {ICartItem} from "../models/ICartItem";

interface CartState {
    cartItems: ICartItem[];
    totalPrice: number;
    isLoading: boolean;
    error: string;
}

const initialState: CartState = {
    cartItems: [],
    totalPrice: 0,
    isLoading: false,
    error: '',
}

export const cartSlice = createSlice({
    name: 'product',
    initialState,
    reducers: {
        addToCart: (state: CartState, action: PayloadAction<ICartItem>) => {
            const totalAmountPrice = action.payload.price * action.payload.amount;
            state.cartItems.push(action.payload);
            state.totalPrice += totalAmountPrice;
        },
        removeFromCart: (state: CartState, action: PayloadAction<{ id: number }>) => {
            const index = state.cartItems.findIndex((item => item.id === action.payload.id));
            state.totalPrice = state.totalPrice - state.cartItems[index].amount * state.cartItems[index].price
            state.cartItems.splice(index, 1);
        },
        increaseAmount: (state: CartState, action: PayloadAction<{ id: number }>) => {
            const item = state.cartItems.find(item => item.id === action.payload.id);
            if (item) {
                item.amount++;
                state.totalPrice += item.price;
            }
        },
        decreaseAmount: (state: CartState, action: PayloadAction<{ id: number }>) => {
            const item = state.cartItems.find(item => item.id === action.payload.id);
            if (item) {
                if (item.amount === 1) {
                    const index = state.cartItems.findIndex((item => item.id === action.payload.id));
                    state.totalPrice = state.totalPrice - state.cartItems[index].amount * state.cartItems[index].price;
                    state.cartItems.splice(index, 1);
                    return;
                }
                item.amount--;
                state.totalPrice -= item.price;
            }
        },
        clearCart: (state: CartState) => {
            state.cartItems.length = 0;
            state.totalPrice = 0;
        }
    }
})

export const {
    addToCart,
    removeFromCart,
    increaseAmount,
    decreaseAmount,
    clearCart,
} = cartSlice.actions;

export default cartSlice.reducer;

