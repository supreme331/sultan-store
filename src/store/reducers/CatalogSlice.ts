import {
    EBodyCare,
    EFaceCare,
    EFootCare,
    EGiftSets,
    EHairCare,
    EHandsCare,
    EShaving,
    ETanning,
    ETypeOfCare
} from "../enums/EProducts";
import {createSlice, PayloadAction} from "@reduxjs/toolkit";
import {IProduct} from "../models/IProduct";
import {initialProducts} from "../initialProducts";
import {fetchProductItems} from "./ActionCreators";

interface CatalogState {
    productItems: IProduct[];
    selectedManufacturers: Array<string>;
    currentTypeOfCare: '' | ETypeOfCare;
    currentSubtypeOfCare: '' | EBodyCare | EHandsCare | EFootCare | EFaceCare | EHairCare | ETanning | EShaving | EGiftSets;
    isLoading: boolean;
    error: string;
}

const initialState: CatalogState = {
    productItems: [],
    selectedManufacturers: [],
    currentTypeOfCare: '',
    currentSubtypeOfCare: '',
    isLoading: true,
    error: '',
}

export const catalogSlice = createSlice({
    name: 'catalog',
    initialState,
    reducers: {
        initializeApp: (state: CatalogState) => {
            state.isLoading = true;
            const data = JSON.stringify(initialProducts);
            localStorage.setItem('productItems', data);
            state.isLoading = false;
        },
        setCurrentTypeOfCare: (state: CatalogState, action: PayloadAction<'' | ETypeOfCare>) => {
            state.currentTypeOfCare = action.payload;
        },
        setCurrentSubtypeOfCare: (state: CatalogState, action: PayloadAction<''
            | EBodyCare
            | EHandsCare
            | EFootCare
            | EFaceCare
            | EHairCare
            | ETanning
            | EShaving
            | EGiftSets>) => {
            state.currentSubtypeOfCare = action.payload;
        },
        showAllProducts: (state: CatalogState) => {
            state.currentTypeOfCare = '';
            state.currentSubtypeOfCare = '';
        },
        showAllProductsOfType: (state: CatalogState, action: PayloadAction<'' | ETypeOfCare>) => {
            state.currentTypeOfCare = action.payload;
            state.currentSubtypeOfCare = '';
        },
        addSelectedManufacturer: (state: CatalogState, action: PayloadAction<string>) => {
            state.selectedManufacturers.push(action.payload)
        },
        deleteSelectedManufacturer: (state: CatalogState, action: PayloadAction<string>) => {
            const index = state.selectedManufacturers.findIndex(item => item === action.payload);
            index !== -1 && state.selectedManufacturers.splice(index, 1);
        },
        clearSelectedManufacturer: (state: CatalogState) => {
            state.selectedManufacturers = [];
        },
        addProductItem: (state: CatalogState, action: PayloadAction<IProduct>) => {
            state.productItems.push(action.payload);
            const data = JSON.stringify(state.productItems);
            localStorage.setItem('productItems', data);
        },
        editProductItem: (state: CatalogState, action: PayloadAction<IProduct>) => {
            const index = state.productItems.findIndex(product => product.id === action.payload.id);
            state.productItems.splice(index, 1, action.payload);
            const data = JSON.stringify(state.productItems);
            localStorage.setItem('productItems', data);
        },
        deleteProductItem: (state: CatalogState, action: PayloadAction<{ productId: number }>) => {
            const index = state.productItems.findIndex(product => product.id === action.payload.productId);
            state.productItems.splice(index, 1);
            const data = JSON.stringify(state.productItems);
            localStorage.setItem('productItems', data);
        },
    },
    extraReducers: {
        [fetchProductItems.pending.type]: (state: CatalogState) => {
            state.productItems = [];
            state.isLoading = true;
        },
        [fetchProductItems.fulfilled.type]: (state: CatalogState, action: PayloadAction<IProduct[]>) => {
            state.productItems = action.payload;
            state.isLoading = false;
        },
        [fetchProductItems.rejected.type]: (state: CatalogState, action: PayloadAction<string>) => {
            state.productItems = [];
            state.isLoading = false;
            state.error = action.payload;
        }
    }
})

export const {
    initializeApp,
    setCurrentTypeOfCare,
    setCurrentSubtypeOfCare,
    showAllProducts,
    showAllProductsOfType,
    addSelectedManufacturer,
    deleteSelectedManufacturer,
    clearSelectedManufacturer,
    addProductItem,
    editProductItem,
    deleteProductItem,
} = catalogSlice.actions;

export default catalogSlice.reducer;