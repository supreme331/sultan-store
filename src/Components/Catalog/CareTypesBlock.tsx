import React from 'react';
import styles from "../../styles/CareTypesBlock.module.scss";
import {showAllProductsOfType} from "../../store/reducers/CatalogSlice";
import {useAppDispatch, useAppSelector} from "../../store/hooks/redux";
import {ETypeOfCare} from "../../store/enums/EProducts";
import {useNavigate} from "react-router";
import {getCurrentTypeOfCare} from "../../store/reducers/selectors/getCurrentTypeOfCare";

const CareTypesBlock = () => {

    const navigate = useNavigate();
    const dispatch = useAppDispatch();
    const currentTypeOfCare = useAppSelector(getCurrentTypeOfCare);

    function onChangeTypeOfCare(value: ETypeOfCare) {
        dispatch(showAllProductsOfType(value));
        navigate('/catalog/1');
    }

    return (
        <ul className={styles.typeOfCare}>
            {
                Object.values(ETypeOfCare)
                    .map((value) =>
                        <li key={value}
                            data-testid={value}
                            className={currentTypeOfCare === value
                                ? styles.typeOfCareActiveItem
                                : styles.typeOfCareItem}
                        onClick={() => onChangeTypeOfCare(value)}>
                            {value}
                        </li>)
            }
        </ul>
    );
};

export default CareTypesBlock;