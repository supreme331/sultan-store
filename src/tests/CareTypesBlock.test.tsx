import {act, render, screen} from '@testing-library/react';
import userEvent from "@testing-library/user-event";
import '@testing-library/jest-dom';
import CareTypesBlock from "../Components/Catalog/CareTypesBlock";
import {MemoryRouter} from "react-router";
import * as reduxHooks from "react-redux";
import * as actions from "../store/reducers/CatalogSlice";

jest.mock('react-redux');

const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');
const mockedShowAllProductsOfType = jest.spyOn(actions, 'showAllProductsOfType');

describe('CareTypesBlock component', () => {
    test('toMatchSnapshot', () => {
        const component = render(
            <MemoryRouter>
                <CareTypesBlock />
            </MemoryRouter>
        );
        expect(component).toMatchSnapshot();
    })

    test('should dispatch actions', () => {
        const dispatch = jest.fn();
        mockedDispatch.mockReturnValue(dispatch);

        render(
            <MemoryRouter>
                <CareTypesBlock />
            </MemoryRouter>
        );

        act(() => {
            userEvent.click(screen.getByTestId('Уход за телом'));
        })

        expect(dispatch).toHaveBeenCalledTimes(1);
        expect(mockedShowAllProductsOfType).toHaveBeenCalledWith('Уход за телом');
    })
})