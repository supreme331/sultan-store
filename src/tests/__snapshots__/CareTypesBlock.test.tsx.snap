// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CareTypesBlock component toMatchSnapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="typeOfCare"
      >
        <li
          class="typeOfCareItem"
          data-testid="Уход за телом"
        >
          Уход за телом
        </li>
        <li
          class="typeOfCareItem"
          data-testid="Уход за руками"
        >
          Уход за руками
        </li>
        <li
          class="typeOfCareItem"
          data-testid="Уход за ногами"
        >
          Уход за ногами
        </li>
        <li
          class="typeOfCareItem"
          data-testid="Уход за лицом"
        >
          Уход за лицом
        </li>
        <li
          class="typeOfCareItem"
          data-testid="Уход за волосами"
        >
          Уход за волосами
        </li>
        <li
          class="typeOfCareItem"
          data-testid="Средства для загара"
        >
          Средства для загара
        </li>
        <li
          class="typeOfCareItem"
          data-testid="Средства для бритья"
        >
          Средства для бритья
        </li>
        <li
          class="typeOfCareItem"
          data-testid="Подарочные наборы"
        >
          Подарочные наборы
        </li>
        <li
          class="typeOfCareItem"
          data-testid="Гигиеническая продукция"
        >
          Гигиеническая продукция
        </li>
        <li
          class="typeOfCareItem"
          data-testid="Гигиена полости рта"
        >
          Гигиена полости рта
        </li>
        <li
          class="typeOfCareItem"
          data-testid="Бумажная продукция"
        >
          Бумажная продукция
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="typeOfCare"
    >
      <li
        class="typeOfCareItem"
        data-testid="Уход за телом"
      >
        Уход за телом
      </li>
      <li
        class="typeOfCareItem"
        data-testid="Уход за руками"
      >
        Уход за руками
      </li>
      <li
        class="typeOfCareItem"
        data-testid="Уход за ногами"
      >
        Уход за ногами
      </li>
      <li
        class="typeOfCareItem"
        data-testid="Уход за лицом"
      >
        Уход за лицом
      </li>
      <li
        class="typeOfCareItem"
        data-testid="Уход за волосами"
      >
        Уход за волосами
      </li>
      <li
        class="typeOfCareItem"
        data-testid="Средства для загара"
      >
        Средства для загара
      </li>
      <li
        class="typeOfCareItem"
        data-testid="Средства для бритья"
      >
        Средства для бритья
      </li>
      <li
        class="typeOfCareItem"
        data-testid="Подарочные наборы"
      >
        Подарочные наборы
      </li>
      <li
        class="typeOfCareItem"
        data-testid="Гигиеническая продукция"
      >
        Гигиеническая продукция
      </li>
      <li
        class="typeOfCareItem"
        data-testid="Гигиена полости рта"
      >
        Гигиена полости рта
      </li>
      <li
        class="typeOfCareItem"
        data-testid="Бумажная продукция"
      >
        Бумажная продукция
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
